va/tlsalpn.go

```diff
@@ -224,9 +249,25 @@ func (va *ValidationAuthorityImpl) validateTLSALPN01(ctx context.Context, identi
 
 	leafCert := certs[0]
 
+	// The certificate may only have the subjectAltName and acmeIdentifier
+	// extensions, and only one of each.
+	allowedOIDs := []asn1.ObjectIdentifier{
+		IdPeAcmeIdentifier, IdCeSubjectAltName,
+	}
+	err := checkAcceptableExtensions(leafCert.Extensions, allowedOIDs)
+	if err != nil {
+		hostPort := net.JoinHostPort(validationRecords[0].AddressUsed.String(), validationRecords[0].Port)
+		errText := fmt.Sprintf(
+			"Incorrect validation certificate for %s challenge. "+
+				"Requested %s from %s. Received %d certificate(s), "+
+				"first certificate had unexpected extensions; got error %s",
+			challenge.Type, identifier.Value, hostPort, len(certs), err)
+		return validationRecords, probs.Unauthorized(errText)
```

Comment:
Not a huge fan of duplicating all of this logic. But the only reasonable way to combine the `checkAcceptableExtensions` and `checkExpectedSAN` checks is to combine all of their functionality into a single helper function. I'm on the fence as to whether combining is a good idea or not.

Reply:
We'll deal with this duplication in some later followup refactoring.